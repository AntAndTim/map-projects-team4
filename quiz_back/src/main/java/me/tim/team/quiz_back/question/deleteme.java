package me.tim.team.quiz_back.question;

public class deleteme {
}
