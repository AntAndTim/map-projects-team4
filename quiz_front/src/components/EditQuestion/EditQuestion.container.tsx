import {QuestionModel} from "../../models/QuestionModel";
import React, {ChangeEvent, useCallback, useState} from "react";
import {message} from "antd";
import {AnswerModel} from "../../models/AnswerModel";

import {CheckboxChangeEvent} from "antd/es/checkbox";
import axios, {AxiosResponse} from "axios";
import {baseUrl} from "../../config/Config";
import {EditQuestion} from "./EditQuestion";

interface EditQuestionContainerProps {
    question?: QuestionModel;
    onSuccess: () => void;
}

export type QuestionTextChaneHandler = (event: ChangeEvent<HTMLInputElement>) => void;
export type AnswerTextChangeHandler = (index: number) => (event: ChangeEvent<HTMLInputElement>) => void;
export type AnswerCorrectnessChangeHandler = (index: number) => (event: CheckboxChangeEvent) => void;
export type DeleteAnswerHandler = (index: number) => () => void;
export type CreateAnswerHandler = () => void;
export type SubmitQuestionHandler = () => void;

export const EditQuestionContainer: React.FC<EditQuestionContainerProps> = ({question, onSuccess}) => {
    const [questionText, setQuestionText] = useState<string>(question?.text || '');
    const [answers, setAnswers] = useState<Partial<AnswerModel>[]>(question?.answers || []);

    const createAnswer = useCallback<CreateAnswerHandler>(() => {
        setAnswers(answers => {
            return [...answers, {text: '', correct: false}];
        });
    }, [setAnswers])

    const submitQuestion = useCallback(() => {
        if (!questionText || !answers || !answers.length || answers.find(({text}) => !Boolean(text))) {
            message.error('Not correct data');

            return;
        }

        if (question?.id) {
            axios.put(baseUrl(`question`), {text: questionText, answers, id: question.id})
                .then(function (response: AxiosResponse) {
                    onSuccess();
                })
                .catch(error => message.error(error.message))
            ;
        } else {
            axios.post(baseUrl(`question`), {text: questionText, answers})
                .then(function (response: AxiosResponse) {
                    onSuccess();
                })
                .catch(error => message.error(error.message))
            ;
        }
    }, [questionText, answers])

    const answerTextChangeHandler = (index: number) =>
        (event: ChangeEvent<HTMLInputElement>) => {
            setAnswers(answers => {
                return answers.map((answer, answerIndex) => {
                    if (index === answerIndex) {
                        return {
                            ...answer,
                            text: event.target.value
                        }
                    }

                    return answer;
                })
            })
        }

    const answerCorrectnessChangeHandler = (index: number) =>
        (event: CheckboxChangeEvent) => {
            setAnswers(answers => {
                return answers.map((answer, answerIndex) => {
                    if (index === answerIndex) {
                        return {
                            ...answer,
                            correct: event.target.checked,
                        }
                    }

                    return answer;
                })
            })
        }

    const deleteAnswer = (index: number) =>
        () => {
            setAnswers(answers => {
                return answers.filter((answer, answerIndex) => index !== answerIndex)
            })
        }

    const questionTextChaneHandler = useCallback<QuestionTextChaneHandler>(event => setQuestionText(event.target.value), [setQuestionText]);

    return (
        <EditQuestion
            questionText={questionText}
            onQuestionTextChange={questionTextChaneHandler}
            answers={answers}
            onAnswerTextChange={answerTextChangeHandler}
            onAnswerCorrectnessChange={answerCorrectnessChangeHandler}
            deleteAnswer={deleteAnswer}
            createAnswer={createAnswer}
            submitQuestion={submitQuestion}
        />
    )
}