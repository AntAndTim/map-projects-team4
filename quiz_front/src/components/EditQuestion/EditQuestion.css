.ant-input.EditQuestion-Text {
    width: 60%;
}

.EditQuestion-Answer {
    margin-bottom: 16px;
}

.EditQuestion-Answer .ant-input {
    width: 50%;
    margin-right: 16px;
}

.ant-checkbox.EditQuestion-AnswerCheckbox {
    margin-right: 16px;
}