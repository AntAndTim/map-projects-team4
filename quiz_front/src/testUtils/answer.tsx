export const getAnswer = (id: number, correct?: boolean, selected?: boolean) => ({
    id,
    text: `AnswerText${id}`,
    correct: Boolean(correct),
    selected: Boolean(selected),
})