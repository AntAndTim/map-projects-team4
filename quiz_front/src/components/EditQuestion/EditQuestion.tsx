import {QuestionModel} from "../../models/QuestionModel";
import React, {ChangeEvent, useCallback, useState} from "react";
import {Button, Checkbox, Divider, Input, message} from "antd";
import {AnswerModel} from "../../models/AnswerModel";

import './EditQuestion.css';
import {CheckboxChangeEvent} from "antd/es/checkbox";
import axios, {AxiosResponse} from "axios";
import {baseUrl} from "../../config/Config";

interface EditQuestionProps {
    question?: QuestionModel
}

export const EditQuestion: React.FC<EditQuestionProps> = ({question}) => {
    const [questionText, setQuestionText] = useState<string>(question?.text || '');
    const [answers, setAnswers] = useState<Partial<AnswerModel>[]>(question?.answers || []);

    const createAnswer = useCallback(() => {
        setAnswers(answers => {
            return [...answers, {text: '', correct: false}];
        });
    }, [setAnswers])

    const submitQuestion = useCallback(() => {
        if (!questionText || !answers || !answers.length || answers.find(({text}) => !Boolean(text))) {
            message.error('Not correct data');

            return;
        }

        if (question?.id) {
            axios.put(baseUrl(`question`), {text: questionText, answers, id: question.id})
                .then(function (response: AxiosResponse) {
                    console.log(response)
                })
                .catch(error => message.error(error.message))
            ;
        } else {
            axios.post(baseUrl(`question`), {text: questionText, answers})
                .then(function (response: AxiosResponse) {
                    console.log(response)
                })
                .catch(error => message.error(error.message))
            ;
        }
    }, [questionText, answers])

    const answerTextChangeHandler = (index: number) =>
        (event: ChangeEvent<HTMLInputElement>) => {
            setAnswers(answers => {
                return answers.map((answer, answerIndex) => {
                    if (index === answerIndex) {
                        return {
                            ...answer,
                            text: event.target.value
                        }
                    }

                    return answer;
                })
            })
        }

    const answerCorrectnessChangeHandler = (index: number) =>
        (event: CheckboxChangeEvent) => {
            setAnswers(answers => {
                return answers.map((answer, answerIndex) => {
                    if (index === answerIndex) {
                        return {
                            ...answer,
                            correct: event.target.checked,
                        }
                    }

                    return answer;
                })
            })
        }

    return (
        <div>
            <Input
                className="EditQuestion-Text"
                placeholder="Question text"
                value={questionText}
                onChange={(event) => setQuestionText(event.target.value)}
            />
            <Divider/>
            {
                answers.map((answer, index) => {
                    return (
                        <div className="EditQuestion-Answer">
                            <Input
                                placeholder="Answer text"
                                value={answer.text}
                                onChange={answerTextChangeHandler(index)}
                            />
                            <span>
                                <Checkbox
                                    className="EditQuestion-AnswerCheckbox"
                                    checked={answer.correct}
                                    onChange={answerCorrectnessChangeHandler(index)}
                                >Correct</Checkbox>
                            </span>
                        </div>
                    )
                })
            }
            <div style={{marginBottom: '16px'}}>
                <Button onClick={createAnswer}>Add answer</Button>
            </div>
            <div>
                <Button onClick={submitQuestion}>Submit</Button>
            </div>
        </div>
    )
}