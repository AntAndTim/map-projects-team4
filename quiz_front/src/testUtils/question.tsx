import {getAnswer} from "./answer";

export const getQuestion = (id: number) => ({
    id,
    text: `QuestionText${id}`,
    answers: [getAnswer(1), getAnswer(2), getAnswer(3, true)],
})