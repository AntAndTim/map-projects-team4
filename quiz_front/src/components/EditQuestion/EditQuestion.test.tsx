import React from 'react';
import {render} from 'enzyme';
import {EditQuestionContainer} from "./EditQuestion.container";
import {getQuestion} from "../../testUtils/question";

const question = getQuestion(100);

describe('EditQuestionContainer', () => {
    it('Should render component', () => {
        const wrapper = render(
            <EditQuestionContainer
                question={question}
                onSuccess={jest.fn}
            />
        );

        expect(wrapper).toMatchSnapshot();
    })
});