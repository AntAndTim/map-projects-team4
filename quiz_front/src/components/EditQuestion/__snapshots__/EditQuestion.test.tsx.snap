// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditQuestionContainer Should render component 1`] = `
<div>
  <input
    class="ant-input EditQuestion-Text"
    placeholder="QuestionContainer text"
    type="text"
    value="QuestionText100"
  />
  <div
    class="ant-divider ant-divider-horizontal"
    role="separator"
  />
  <div
    class="EditQuestion-Answer"
  >
    <input
      class="ant-input"
      placeholder="Answer text"
      type="text"
      value="AnswerText1"
    />
    <span>
      <label
        class="ant-checkbox-wrapper EditQuestion-AnswerCheckbox"
      >
        <span
          class="ant-checkbox"
        >
          <input
            class="ant-checkbox-input"
            type="checkbox"
          />
          <span
            class="ant-checkbox-inner"
          />
        </span>
        <span>
          Correct
        </span>
      </label>
    </span>
    <button
      class="ant-btn"
      type="button"
    >
      <span>
        Delete
      </span>
    </button>
  </div>
  <div
    class="EditQuestion-Answer"
  >
    <input
      class="ant-input"
      placeholder="Answer text"
      type="text"
      value="AnswerText2"
    />
    <span>
      <label
        class="ant-checkbox-wrapper EditQuestion-AnswerCheckbox"
      >
        <span
          class="ant-checkbox"
        >
          <input
            class="ant-checkbox-input"
            type="checkbox"
          />
          <span
            class="ant-checkbox-inner"
          />
        </span>
        <span>
          Correct
        </span>
      </label>
    </span>
    <button
      class="ant-btn"
      type="button"
    >
      <span>
        Delete
      </span>
    </button>
  </div>
  <div
    class="EditQuestion-Answer"
  >
    <input
      class="ant-input"
      placeholder="Answer text"
      type="text"
      value="AnswerText3"
    />
    <span>
      <label
        class="ant-checkbox-wrapper ant-checkbox-wrapper-checked EditQuestion-AnswerCheckbox"
      >
        <span
          class="ant-checkbox ant-checkbox-checked"
        >
          <input
            checked=""
            class="ant-checkbox-input"
            type="checkbox"
          />
          <span
            class="ant-checkbox-inner"
          />
        </span>
        <span>
          Correct
        </span>
      </label>
    </span>
    <button
      class="ant-btn"
      type="button"
    >
      <span>
        Delete
      </span>
    </button>
  </div>
  <div
    style="margin-bottom:16px"
  >
    <button
      class="ant-btn"
      type="button"
    >
      <span>
        Add answer
      </span>
    </button>
  </div>
  <div>
    <button
      class="ant-btn"
      type="button"
    >
      <span>
        Submit
      </span>
    </button>
  </div>
</div>
`;
